apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1.Tests/SubscriberClientTest.cs

```diff
@@ -1470,5 +1481,178 @@ public void ExactlyOnceDelivery_Extends_MixedFault()
+                await fake.TaskHelper.ConfigureAwait(fake.Subscriber.StopAsync(CancellationToken.None));
+                // All 4 messages are handled.
+                Assert.Equal(4, handledMsgs.Count);
+                Assert.Equal(new[] { "1", "2", "3", "4" }, handledMsgs);
+            });
+        }
+
+        [Fact]
+        public void ExactlyOnceDelivery_ReceiptModAck_MixedFault()
+        {
+            var msgs = new[]
+            {
+                ServerAction.Data(TimeSpan.Zero, new[] { "1" }),
+                ServerAction.Data(TimeSpan.Zero, new[] { "3" }),
+                ServerAction.Data(TimeSpan.Zero, new[] { "2" }),
+                ServerAction.Data(TimeSpan.Zero, new[] { "4" }),
+                ServerAction.Inf()
+            };
+
+            // Message 1,4 is success, 2 is temporary failure and 3 is permanent failure.
+            var exception = GetExactlyOnceDeliveryMixedException();
+            // We have both temporary and permanent failures.
+            // Temporary failure in extend request should be retried.
+            var ackModifyAckDeadlineAction = AckModifyAckDeadlineAction.BadExtend(exception, numberOfFailures: 2);
+
+            using var fake = Fake.Create(new[] { msgs }, useMsgAsId: true, ackModifyAckDeadlineAction: ackModifyAckDeadlineAction, isExactlyOnceDelivery: true);
+
+            fake.Scheduler.Run(async () =>
+            {
+                var handledMsgs = new List<string>();
+                var doneTask = fake.Subscriber.StartAsync(async (msg, ct) =>
+                {
+                    handledMsgs.Locked(() => handledMsgs.Add(msg.Data.ToStringUtf8()));
+                    await fake.TaskHelper.ConfigureAwait(fake.Scheduler.Delay(TimeSpan.FromSeconds(10), ct));
+                    return await Task.FromResult(SubscriberClient.Reply.Ack);
+                });
+
+                await fake.TaskHelper.ConfigureAwait(fake.Scheduler.Delay(TimeSpan.FromSeconds(100), CancellationToken.None));
+                await fake.TaskHelper.ConfigureAwait(fake.Subscriber.StopAsync(CancellationToken.None));
+                // Permanently failed receipt ModAcks won't be passed to the user handler, so 3 is not handled.
+                // We have simulated the scenario that temporary failed ModAck (message 2) becomes successful after 1 retry and hence is handled.
+                Assert.Equal(new[] { "1", "2", "4" }, handledMsgs);
+            });
+        }
+
+        [Fact]
+        public void ExactlyOnceDelivery_ReceiptModAck_PermanentFaults()
+        {
+            var msgs = new[]
+            {
+                ServerAction.Data(TimeSpan.Zero, new[] { "1" }),
+                ServerAction.Data(TimeSpan.Zero, new[] { "2" }),
+                ServerAction.Data(TimeSpan.Zero, new[] { "3" }),
+                ServerAction.Data(TimeSpan.Zero, new[] { "4" }),
+                ServerAction.Inf()
+            };
+
+            // Message 1,2,3,4 are all permanent failures.
```

Comment:
So why are we calling GetExactlyOnceDeliveryMixedException? I thought that would make 2 a temporary failure, and 1 and 4 success.

Reply:
If we use `GetExactlyOnceDeliveryMixedException` without the `ErrorInfo` parameter it would make 2 a temporary failure, 1 and 4 as success as a default implementation. Here, I have passed the `ErrorInfo` to make all of the messages as permanent failures.